luals-addon/factorio/factorio-plugin/util.lua

```diff
@@ -368,6 +374,221 @@ local function add_chain_diff(chain_diff, diffs)
   end
 end
 
+---@alias LexerState
+--- | '"code"'
+--- | '"short_string"'
+--- | '"long_string"'
+--- | '"short_comment"'
+--- | '"long_comment"'
+
+---@type integer[]
+local code_ranges = {0} -- Always contains 1 element.
+local code_ranges_count = 1
+local current_code_ranges_lower_bound = 0 -- Zero based.
+
+function clean_up_code_ranges()
+  for i = code_ranges_count, 2, -1 do
+    code_ranges[i] = nil
+  end
+  code_ranges_count = 1
+  current_code_ranges_lower_bound = 0
+end
+
+local function reset_code_ranges()
+  current_code_ranges_lower_bound = 0
+end
+
+---Lexically analyze Lua source files for positions of strings and comments.
+---Notably, this needs to be able to handle 'long brackets', which are context-sensitive.
+---We should really only be doing this once per source file.
+---@param source string
+function lex_lua_nonexecutables(source)
+  ---@type LexerState
+  local state = "code"
+  local cursor = 1 -- 1 is the first character in the source file.
+
+  local delimit = ""
+  local patterned_delimit = ""
+  local start = 0
+  local _end = 0
+
+  local char_escaped = false
+
+  ---check if the next character(s) are equal to the given string
+  ---@param query string
+  ---@return boolean
+  local function peek(query)
+    if cursor + #query > #source then return false end
+    return source:sub(cursor, cursor + #query - 1) == query
+  end
+
+  local function take(query)
+    if not peek(query) then return false end
+    cursor = cursor + #query
+    return true
+  end
+
+  ---Parse a opening long bracket, like `[[` or `[=[`
+  ---Assumes the first bracket has already been consumed.
+  ---@return boolean, integer | nil
+  local function parse_longbracket_open()
+    -- Consume all the '='s
+    local count = source:match("^=*()", cursor) - cursor
+    cursor = cursor + count
+    if take("[") then
+      return true, count
+    else
+      return false, nil
+    end
```

Comment:
Yea it would, would be better tbh. Though something I have tested in the past is that having a match return a string vs retunring a number is slower, so I'd keep the `()` at the end to calculate count and to update cursor in any way that works.
Although saying that I did test performance before and after moving the modes switch table outside of the loop and the difference was within margin of error (so there was no difference), so realistically the suggested change is good as is.